numbersofinterest: developing numbers of interest algorithm in TDD

aurelienlair opened the pull request:
This pull request is work in progress.
Here is the description of the problem.

Harold always boasted about his prowess with numbers. So one day Reese challenged him to a problem. He gave Harold two numbers X and Y and asked him to find out the Nth number of the series which began with X numbers of Y’s and the following elements are equal to the sum of the last X numbers in the series. Help Harold find the Nth number of the series.

Input:
The first line contains a number T , number of test cases.
The next T lines each contain 3 integers X, Y and N separated by single space.

Output: 
For each test case print the Nth number of the sequence.

Constraints:
1<=T<=10^6
0 < N <= 50
0 < X < 50
0 <= Y< 3

For more details please check:
https://www.hackerearth.com/problem/algorithm/numbers-of-interest-1/

Commit:
numbersofinterest: developing numbers of interest algorithm in TDD

Commit:
adding space

Commit:
adding test on invalid number of series

giorgiosironi commented on numbersofinterest/NumbersOfInterest.php:
```diff
@@ -0,0 +1,58 @@
+
+class NumbersOfInterest
+{
+    public static function calculate($input)
```
would investigate turning input into a data structure (e.g. `class Input`) so that parsing is separated from its usage

giorgiosironi commented on numbersofinterest/NumbersOfInterest.php:
```diff
@@ -0,0 +1,58 @@
+
+        for ($cursor=0; $cursor  < ($n-$x); $cursor++) {
+            $sum = 0;
+            for ($z=0; $z<$x; $z++) {
```
would try to extract the second for into a private method, investigate what parameters could be passed

giorgiosironi replied:
$currentSeries[$z+$cursor] = self::sum(...)

giorgiosironi commented on numbersofinterest/NumbersOfInterest.php:
```diff
@@ -0,0 +1,58 @@
+
+class SeriesArrayToConsoleLine
+{
+    public static function output(array $series)
```
readable, wonder why only `static` methods

giorgiosironi commented on numbersofinterest/NumbersOfInterest.php:
```diff
@@ -0,0 +1,58 @@
+        return implode(
+            PHP_EOL,
+            array_map(
+                function($element)
```
maybe `$series`

giorgiosironi commented on numbersofinterest/NumbersOfInterestTest.php:
```diff
@@ -0,0 +1,72 @@
+        $input = "$numberOfSeries" . PHP_EOL . "$x $y $n"; 
+
+        $this->assertEquals(
+            '2, 2, 2, 6, 10, 18, 34',
```
separate testing output format from testing the solution is correct

Commit:
refactoring as suggested by @giorgiosironi

giorgiosironi commented on numbersofinterest/Algebraist.php:
```diff
@@ -0,0 +1,47 @@
+use NumbersOfInterest\Options;
+use NumbersOfInterest\SeriesProperties;
+
+class Algebraist
```
of for the abstraction name, quite uncommon but correct

giorgiosironi replied:
with this abstraction I believe all methods should become non-static

giorgiosironi commented on numbersofinterest/Algebraist.php:
```diff
@@ -0,0 +1,47 @@
+require_once __DIR__ . '/SeriesArrayToConsoleLine.php';
+require_once __DIR__ . '/Options.php';
+
+use NumbersOfInterest\Options;
```
you don't need this `use` statements, you are already inside the namespace

giorgiosironi commented on numbersofinterest/Algebraist.php:
```diff
@@ -0,0 +1,47 @@
+
+        for ($cursor=0; $cursor< ($properties->totalOfNumbers-$properties->occurencesOfReferenceNumber); $cursor++) {
+            // adding sum at position cursor + occurencesOfReferenceNumber
+            $currentSeries[$cursor+$properties->occurencesOfReferenceNumber] = self::sumUntil($properties->occurencesOfReferenceNumber, $currentSeries, $cursor);
```
improved, shorter and more abstract method

giorgiosironi commented on numbersofinterest/Options.php:
```diff
@@ -0,0 +1,33 @@
+        $this->seriesOptions = $raws; 
+    }
+
+    public function toArray()
```
`allSeries()`? `all()`, referring to the options? it's an array of what?

giorgiosironi commented on numbersofinterest/OptionsTest.php:
```diff
@@ -0,0 +1,56 @@
+require_once __DIR__ . '/Options.php';
+use NumbersOfInterest\Options;
+
+class OptionsTest extends \PHPUnit_Framework_TestCase
```
nice to test this in isolation, also with the corner cases such as invalid input

giorgiosironi commented on numbersofinterest/SeriesPropertiesTest.php:
```diff
@@ -0,0 +1,69 @@
+    /**
+     * @expectedException InvalidArgumentException
+     */
+    public function testWhenSeriesPropertiesIsNotParsableWithSeparatorAnExceptionIsThrown()
```
someone could say that the solution is overengineered due to the higher number of classes; I would try to organize them in folders separating:
- tests from production code
- domain from user interface (input and output management)

Commit:
Reorganizing object structures by dividing domain and ui.
Replacing static function by instance functions as suggested by
@giorgiosironi
Divide code and tests

Commit:
correcting function all on test
